libraries/standard/http/src/http_client.c

```diff
@@ -725,68 +801,135 @@ static HTTPStatus_t sendHttpHeaders( const HTTPTransportInterface_t * pTransport
+                "HeaderBytes=%d",
+                headerLength ) );
+
+    returnStatus = sendHttpData( pTransport, ( const uint8_t * ) pContentLengthBuffer, headerLength );
+
+    return returnStatus;
+}
+
+/*-----------------------------------------------------------*/
+
+static HTTPStatus_t sendHttpHeaders( const HTTPTransportInterface_t * pTransport,
+                                     const HTTPRequestHeaders_t * pRequestHeaders,
+                                     size_t reqBodyBufLen,
+                                     uint32_t flags )
+{
+    HTTPStatus_t returnStatus = HTTP_SUCCESS;
+    size_t numBytesToSend = 0u;
+    uint8_t shouldSendContentLength = 0u;
 
     assert( pTransport != NULL );
     assert( pTransport->send != NULL );
-    assert( pRequestBodyBuf != NULL );
+    assert( pRequestHeaders != NULL );
 
-    transportStatus = pTransport->send( pTransport->pContext,
-                                        pRequestBodyBuf,
-                                        reqBodyBufLen );
+    numBytesToSend = pRequestHeaders->headersLen;
 
-    if( transportStatus < 0 )
-    {
-        LogError( ( "Failed to send HTTP body: Transport send() "
-                    " returned error: TransportStatus=%d",
-                    transportStatus ) );
-        returnStatus = HTTP_NETWORK_ERROR;
-    }
-    else if( ( size_t ) transportStatus != reqBodyBufLen )
+    /* Send the content length header if the flag to disable is not set and the
+     * body length is greater than zero. */
+    shouldSendContentLength = ( ( ( flags & HTTP_SEND_DISABLE_CONTENT_LENGTH_FLAG ) == 0u ) &&
+                                ( reqBodyBufLen > 0u ) ) ? 1u : 0u;
+
+    /* If the application does want the Content-Length header automatically
+     * written. Then send the headers in pRequestHeaders without the final
+     * "\r\n" that denotes the end of the header. */
+    if( shouldSendContentLength == 1u )
     {
-        LogError( ( "Failed to send HTTP body: Transport send() "
-                    "did not send the required bytes: RequiredBytes=%u"
-                    ", SentBytes=%d.",
-                    reqBodyBufLen,
-                    transportStatus ) );
-        returnStatus = HTTP_NETWORK_ERROR;
+        numBytesToSend -= HTTP_HEADER_LINE_SEPARATOR_LEN;
     }
-    else
+
+    LogDebug( ( "Sending HTTP request headers: HeaderBytes=%d",
+                numBytesToSend ) );
+
+    /* Send the HTTP headers over the network. */
+    returnStatus = sendHttpData( pTransport, pRequestHeaders->pBuffer, numBytesToSend );
+
+    if( returnStatus == HTTP_SUCCESS )
     {
-        LogDebug( ( "Sent HTTP body over the transport: BytesSent=%d.",
-                    transportStatus ) );
+        if( shouldSendContentLength == 1u )
+        {
+            /* Send the Content-Length header over the network. */
+            returnStatus = sendHttpContentLength( pTransport, reqBodyBufLen );
```

Comment:
What is the reason for sending content length separately and not writing in the headers buffer and send all the headers in one go?

Reply:
Then the customer would need to plan more carefully the buffer space needed for their header buffer. This is to help the customer experience, what do you think?

Reply:
Offline discussion we have decided that it is better to save resources that could be used by invoking transport_write too many times. We shall have the application make sure there is enough space in the request header buffer for the automatically written content-length.